src/librustc/middle/infer/sub.rs

```diff
@@ -86,6 +86,7 @@ impl<'a, 'tcx> TypeRelation<'a, 'tcx> for Sub<'a, 'tcx> {
             }
 
             (&ty::TyError, _) | (_, &ty::TyError) => {
+                infcx.set_tainted_by_errors();
```

Comment:
@arielb1 

> "causing damage"

Yes, we could remove it. I don't think it's needed -- rather, we should set the flag at the *source* of the errors, as I later did. (I could also assert when setting the flag that we are not in a snapshot, just to be safe.)